pkg/internal/itest/testpeer/generator.go

```diff
@@ -259,3 +329,89 @@ func StartAndWaitForReady(ctx context.Context, manager datatransfer.Manager) err
 		return err
 	}
 }
+
+func HttpHandler(ctx context.Context, lsys linking.LinkSystem) func(http.ResponseWriter, *http.Request) {
+	return func(res http.ResponseWriter, req *http.Request) {
+		urlPath := strings.Split(req.URL.Path, "/")[1:]
+
+		// validate CID path parameter
+		cidStr := urlPath[1]
+		rootCid, err := cid.Parse(cidStr)
+		if err != nil {
+			http.Error(res, fmt.Sprintf("Failed to parse CID path parameter: %s", cidStr), http.StatusBadRequest)
+			return
+		}
+
+		// Grab unixfs path if it exists
+		unixfsPath := ""
+		if len(urlPath) > 2 {
+			unixfsPath = "/" + strings.Join(urlPath[2:], "/")
+		}
+
+		// Parse car scope and use it to get selector
+		carScope := types.CarScopeAll
+		if req.URL.Query().Has("car-scope") {
+			switch req.URL.Query().Get("car-scope") {
+			case "all":
+			case "file":
+				carScope = types.CarScopeFile
+			case "block":
+				carScope = types.CarScopeBlock
+			default:
+				http.Error(res, fmt.Sprintf("Invalid  car-scope parameter: %s", req.URL.Query().Get("car-scope")), http.StatusBadRequest)
+				return
+			}
+		}
+		selNode := unixfsnode.UnixFSPathSelectorBuilder(unixfsPath, carScope.TerminalSelectorSpec(), false)
+		sel, err := selector.CompileSelector(selNode)
+		if err != nil {
+			http.Error(res, fmt.Sprintf("Failed to compile selector from car-scope: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		// Write to response writer
+		carWriter, err := storage.NewWritable(res, []cid.Cid{rootCid}, car.WriteAsCarV1(true), car.AllowDuplicatePuts(false))
+		if err != nil {
+			http.Error(res, fmt.Sprintf("Failed to create car writer: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		// Extend the StorageReadOpener func to write to the carWriter
+		originalSRO := lsys.StorageReadOpener
+		lsys.StorageReadOpener = func(lc linking.LinkContext, lnk datamodel.Link) (io.Reader, error) {
+			r, err := originalSRO(lc, lnk)
+			if err != nil {
+				return nil, err
+			}
+			byts, err := io.ReadAll(r)
+			if err != nil {
+				return nil, err
+			}
+			err = carWriter.Put(ctx, lnk.(cidlink.Link).Cid.KeyString(), byts)
+			if err != nil {
+				return nil, err
+			}
+
+			return bytes.NewReader(byts), nil
+		}
+
+		rootNode, err := lsys.Load(linking.LinkContext{}, cidlink.Link{Cid: rootCid}, basicnode.Prototype.Any)
+		if err != nil {
+			http.Error(res, fmt.Sprintf("Failed to load root cid into link system: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		cfg := &traversal.Config{
+			Ctx:                            ctx,
+			LinkSystem:                     lsys,
+			LinkTargetNodePrototypeChooser: basicnode.Chooser,
+		}
+		progress := traversal.Progress{Cfg: cfg}
+
+		err = progress.WalkAdv(rootNode, sel, func(p traversal.Progress, n datamodel.Node, vr traversal.VisitReason) error { return nil })
```

Comment:
I'm finding that pulling the visitor function out as a static in your test file makes the code cleaner, so you can:

```suggestion
		err = progress.WalkAdv(rootNode, sel, visitNoop)
```

with

```go
func visitNoop(p traversal.Progress, n datamodel.Node, vr traversal.VisitReason) error { return nil }
```

Reply:
@rvagg I hope for the day when the signature is progress.Walk(rootNode, sel) with visitor fn being a config on the traversal :P